### ISSUE TITLE
vite build error

### ISSUE THREAD

**Issue by hellhorse123** [epam/ketcher#5565]
Steps to reproduce:

1. npm create vite@latest
2. choose React, typescript
3.  install libs "ketcher-react": "^2.24.0", "ketcher-standalone": "^2.24.0",
4.  invoke libs in component, for example: 
```
/* eslint-disable @typescript-eslint/no-explicit-any */
import { useEffect, useState } from "react";
import { Ketcher, StandaloneStructServiceProvider } from "ketcher-standalone";
import { Editor } from "ketcher-react";
import "ketcher-react/dist/index.css";
import "./App.css";

const structServiceProvider = new StandaloneStructServiceProvider();


function App() {
  const [isKetcherInitialized, setIsKetcherInitialized] = useState(false); // Track Ketcher initialization
  const [ketcherInstance, setKetcherInstance] = useState<Ketcher | null>(null); // Store Ketcher instance

  useEffect(() => {
    if (ketcherInstance && isKetcherInitialized) {
      try {
        ketcherInstance.setMolecule("");
      } catch (error) {
        console.error("Failed to set SMILES:", error);
      }
    }
  }, [isKetcherInitialized]);

  return (
    <div>
      <Editor
        staticResourcesUrl={""}
        structServiceProvider={structServiceProvider as any}
        errorHandler={(message: string | Error) => {
          console.error("Error in Ketcher:", message);
      
        }}
        onInit={(ketcher: Ketcher) => {
          // Assign Ketcher instance to window for KetcherLogger to use

          /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
          // @ts-ignore
          window.ketcher = ketcher;

          setKetcherInstance(ketcher); // Set Ketcher instance in state
          setIsKetcherInitialized(true); // Mark Ketcher as initialized
          try {
            ketcher.setMolecule(''); // Set the default SMILES after initialization
          } catch (error) {
            console.error("Error initializing Ketcher with SMILES:", error);
          }
        }}
      />
    </div>
  );
}

export default App;
```
5. Try to build: npm run build

Explanation:
After adding libs:
```
    "ketcher-react": "^2.24.0",
    "ketcher-standalone": "^2.24.0",
```

and trying to build my react ts project with vite, I get error:
```
thread '<unnamed>' panicked at /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/swc_common-0.39.0/src/syntax_pos.rs:702:18:
width 3 given for non-narrow character
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at library/core/src/panicking.rs:221:5:
panic in a function that cannot unwind
stack backtrace:
   0:        0x1329fce37 - _napi_register_module_v1
   1:        0x1329899eb - <unknown>
   2:        0x1329d19f2 - _napi_register_module_v1
   3:        0x1329fe442 - _napi_register_module_v1
   4:        0x1329ff8d9 - _napi_register_module_v1
   5:        0x1329fe992 - _napi_register_module_v1
   6:        0x1329fe929 - _napi_register_module_v1
   7:        0x1329fe91c - _napi_register_module_v1
   8:        0x132b4c56c - _napi_register_module_v1
   9:        0x132b4c5e5 - _napi_register_module_v1
  10:        0x132b4c589 - _napi_register_module_v1
  11:        0x132981dfc - <unknown>
  12:        0x100616b40 - __ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
thread caused non-unwinding panic. aborting.
/bin/sh: line 1: 13941 Abort trap: 6           vite build
```

As I understand, this is trouble : `width 3 given for non-narrow character`, chatGPT says:

> SWC is a Rust-based compiler that Vite uses for code transformation and minification during the build process. The error message:
> width 3 given for non-narrow character
> indicates that SWC is encountering a character that doesn't fit within its expected character width, possibly due to Unicode characters or certain syntax in the ketcher packages.
> 
> Given that the error appears after importing ketcher-react and ketcher-standalone, it's likely that:
> 
>     SWC cannot process some code within these packages.
>     The packages might contain non-standard syntax or characters that cause SWC to crash.

my dependencies:
```
    "react": "^18.2.0",
    "vite": "^5.3.2",
```

**Comment by hellhorse123** [comment]
More precisely, there is a bug in the `ketcher-react` library, without it the build is successful

**Comment by hellhorse123** [comment]
One more addition:
Here is the problematic line:
![telegram-cloud-photo-size-2-5449784376289254798-y](https://github.com/user-attachments/assets/0655bf09-436b-42c1-933a-9c0654223789)

[Code file with error](https://github.com/epam/ketcher/blob/95cf1635c06da148d0eca5ee4b8b52f18eed6c11/packages/ketcher-core/src/application/render/raphael-ext.ts#L26)

After the build, for some reason this import does not work.

![telegram-cloud-photo-size-2-5449784376289254813-x](https://github.com/user-attachments/assets/0ec2048e-b902-49a8-b96f-d9b98c988ecb)
![telegram-cloud-photo-size-2-5449784376289254814-y](https://github.com/user-attachments/assets/5323dc6c-22dd-4b1e-ac14-8f88b73b956f)
![telegram-cloud-photo-size-2-5449784376289254817-y](https://github.com/user-attachments/assets/9e2bcffa-697c-4401-a0d2-2719775bcd7b)

**Comment by samellis** [comment]
I get the same bug, also vite, no errors apparently in build but trying to load site throws this error. npm run start serves a working page, problem is only after build.

> After adding libs:
> 
> ```
>     "ketcher-react": "^2.24.0",
>     "ketcher-standalone": "^2.24.0",
> ```
> 
> and trying to build my react ts project with vite, I get error:
> 
> ```
> thread '<unnamed>' panicked at /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/swc_common-0.39.0/src/syntax_pos.rs:702:18:
> width 3 given for non-narrow character
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> thread '<unnamed>' panicked at library/core/src/panicking.rs:221:5:
> panic in a function that cannot unwind
> stack backtrace:
>    0:        0x1329fce37 - _napi_register_module_v1
>    1:        0x1329899eb - <unknown>
>    2:        0x1329d19f2 - _napi_register_module_v1
>    3:        0x1329fe442 - _napi_register_module_v1
>    4:        0x1329ff8d9 - _napi_register_module_v1
>    5:        0x1329fe992 - _napi_register_module_v1
>    6:        0x1329fe929 - _napi_register_module_v1
>    7:        0x1329fe91c - _napi_register_module_v1
>    8:        0x132b4c56c - _napi_register_module_v1
>    9:        0x132b4c5e5 - _napi_register_module_v1
>   10:        0x132b4c589 - _napi_register_module_v1
>   11:        0x132981dfc - <unknown>
>   12:        0x100616b40 - __ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
> thread caused non-unwinding panic. aborting.
> /bin/sh: line 1: 13941 Abort trap: 6           vite build
> ```
> 
> As I understand, this is trouble : `width 3 given for non-narrow character`, chatGPT says:
> 
> > SWC is a Rust-based compiler that Vite uses for code transformation and minification during the build process. The error message:
> > width 3 given for non-narrow character
> > indicates that SWC is encountering a character that doesn't fit within its expected character width, possibly due to Unicode characters or certain syntax in the ketcher packages.
> > Given that the error appears after importing ketcher-react and ketcher-standalone, it's likely that:
> > ```
> > SWC cannot process some code within these packages.
> > The packages might contain non-standard syntax or characters that cause SWC to crash.
> > ```
> 
> my dependencies:
> 
> ```
>     "react": "^18.2.0",
>     "vite": "^5.3.2",
> ```

I had this problem, found the solution in this vite bug: https://github.com/vitejs/vite/issues/18179#issue-2543316736
```
npm install -D rollup@4.22.0
```

**Comment by samellis** [comment]
> One more addition: Here is the problematic line: ![telegram-cloud-photo-size-2-5449784376289254798-y](https://private-user-images.githubusercontent.com/48678379/370637163-0655bf09-436b-42c1-933a-9c0654223789.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNjg0MzYsIm5iZiI6MTcyNzM2ODEzNiwicGF0aCI6Ii80ODY3ODM3OS8zNzA2MzcxNjMtMDY1NWJmMDktNDM2Yi00MmMxLTkzM2EtOWMwNjU0MjIzNzg5LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDE2Mjg1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjZmFiZjc3NDY0MzgyOGE0NmNkNWYzZGE3NjhmNDVmZDFmNWEwYTIzYTZkM2E2NTU4NzJlOWNkMDIzNmYyNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yva6msoIBfqTuYucJk71QEj4IgIWhT81pp6hEFFB1h0)
> 
> [Code file with error](https://github.com/epam/ketcher/blob/95cf1635c06da148d0eca5ee4b8b52f18eed6c11/packages/ketcher-core/src/application/render/raphael-ext.ts#L26)
> 
> After the build, for some reason this import does not work.
> 
> ![telegram-cloud-photo-size-2-5449784376289254813-x](https://private-user-images.githubusercontent.com/48678379/370637224-0ec2048e-b902-49a8-b96f-d9b98c988ecb.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNjg0MzYsIm5iZiI6MTcyNzM2ODEzNiwicGF0aCI6Ii80ODY3ODM3OS8zNzA2MzcyMjQtMGVjMjA0OGUtYjkwMi00OWE4LWI5NmYtZDliOThjOTg4ZWNiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDE2Mjg1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3NjcxOGVhNTNiNTVjZWFjODkwNTJjYjc0OTNmNzg5Njg5OWY5ZjAyNGYwZDZkOTAyNmM1YzQ1Y2E2MDViMTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Gh-pveDNVD2u6vw5w7aqMR7kBdFef93qsURQrhWPjnw) ![telegram-cloud-photo-size-2-5449784376289254814-y](https://private-user-images.githubusercontent.com/48678379/370637261-5323dc6c-22dd-4b1e-ac14-8f88b73b956f.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNjg0MzYsIm5iZiI6MTcyNzM2ODEzNiwicGF0aCI6Ii80ODY3ODM3OS8zNzA2MzcyNjEtNTMyM2RjNmMtMjJkZC00YjFlLWFjMTQtOGY4OGI3M2I5NTZmLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDE2Mjg1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMDhiOTQwNDA4ODJiYjQ2NzNjY2NmMTA0NmQ3NjQxYjczZjkyZWY5NDBkNTg1NWYxM2EwNzdkMWMwMDQwMWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.eSwZ8wkAG2i_ZKoRgRGaw5g8c9Sn9RBm-UBMJcENwXI) ![telegram-cloud-photo-size-2-5449784376289254817-y](https://private-user-images.githubusercontent.com/48678379/370637305-9e2bcffa-697c-4401-a0d2-2719775bcd7b.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNjg0MzYsIm5iZiI6MTcyNzM2ODEzNiwicGF0aCI6Ii80ODY3ODM3OS8zNzA2MzczMDUtOWUyYmNmZmEtNjk3Yy00NDAxLWEwZDItMjcxOTc3NWJjZDdiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDE2Mjg1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiMzIyNWQ4ZGFiYTkxZDcwOTMwNDgzNWFiN2E0YzAwMjViNTI5NDdjNWQ4NTI0NTJjNGM0NjZmNjFlMWZjOTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9RwOsb7hekcc8iDTQwgyEIaSaK7YSWGRn7CE197-I90)

After solving the first post issue (with npm install -D rollup@4.22.0) I managed to build successfully. However on opening the site I see the same "Uncaught ReferenceError: require is not defined" on the same bit of code (raphael) in ketcher-react.

**Comment by wiscgazf** [comment]
May I ask, after vite packaging, the problem of require('raphael') require not defined has been solved? I have exactly the same phenomenon as you

![微信图片_20241011135735](https://github.com/user-attachments/assets/9a96fa94-75c6-478a-a042-9ab1398b4fad)
![微信图片_20241011135745](https://github.com/user-attachments/assets/9f51227d-b361-4ad3-87c8-b793458b444b)
![微信图片_20241011135754](https://github.com/user-attachments/assets/edc83956-b843-4ebd-aef1-fa2a9c158d35)

**Comment by hellhorse123** [comment]
> May I ask, after vite packaging, the problem of require('raphael') require not defined has been solved? I have exactly the same phenomenon as you
> 
> ![微信图片_20241011135735](https://private-user-images.githubusercontent.com/26874662/375641146-9a96fa94-75c6-478a-a042-9ab1398b4fad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MzYzNTEsIm5iZiI6MTcyODYzNjA1MSwicGF0aCI6Ii8yNjg3NDY2Mi8zNzU2NDExNDYtOWE5NmZhOTQtNzVjNi00NzhhLWEwNDItOWFiMTM5OGI0ZmFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDA4NDA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0Y2MxYmVmYzE0ZjY0NjI1MGVmOWEyMWQwODUyNjU2ODRmMzJhNTZlOTU5Y2U1MDE1NTNkZDI4YWNiMDQwMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HZn_HYc-yJLzHQ_cYS19YymKzcFHtMnNdE1rcabQVyw) ![微信图片_20241011135745](https://private-user-images.githubusercontent.com/26874662/375641162-9f51227d-b361-4ad3-87c8-b793458b444b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MzYzNTEsIm5iZiI6MTcyODYzNjA1MSwicGF0aCI6Ii8yNjg3NDY2Mi8zNzU2NDExNjItOWY1MTIyN2QtYjM2MS00YWQzLTg3YzgtYjc5MzQ1OGI0NDRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDA4NDA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0Y2M2OTEwNDk3ZDc4MDYxMjRiN2Q2YTg5ZWNmNDM4MmQ4NDM3NDk4NmE1NDRiZjI0YTIxODhjNzY2N2JiOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K7A-UwOpwyig_-y8eZOrZ9cnY_KACIG77AbTxTyGU7Q) ![微信图片_20241011135754](https://private-user-images.githubusercontent.com/26874662/375641168-edc83956-b843-4ebd-aef1-fa2a9c158d35.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MzYzNTEsIm5iZiI6MTcyODYzNjA1MSwicGF0aCI6Ii8yNjg3NDY2Mi8zNzU2NDExNjgtZWRjODM5NTYtYjg0My00ZWJkLWFlZjEtZmEyYTljMTU4ZDM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDA4NDA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiOTUyODJiZDY1YzBmYTQzN2QzOGFhNjFkOGJjYTM1MzA2N2ZjZDBkYzE3NDJkZGFjYjIxYzhlMzA4ZTM5YTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.R8zrOlcDXCkCytZI5LLn7xSnQuwy_kkOftbYGh1Cpt8)

Unfortunately, the problem is still not solved and there is no feedback from the developers

**Comment by wiscgazf** [comment]
`
yarn add vite-plugin-commonjs

import commonjs from 'vite-plugin-commonjs';

// vite.config
{
plugins: [react(), commonjs()],
}
`

**Comment by wiscgazf** [comment]
``` 
I have solved the problem ‘the problem of require('raphael') require not defined ’


// 1、 yarn add vite-plugin-commonjs

// vite.config.ts

// 2、import commonjs from 'vite-plugin-commonjs';

{
plugins: [react(), commonjs()],
define: {
    'process.env': {},
    global: 'global',
  },
  build: {
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },
}

// 3、index.html
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + React + TS</title>
         // add script code add to head
    <script>
      var global = window;
    </script>
  </head>
```

**Comment by hellhorse123** [comment]
> 

in my case, installing rollup (or overriding it) does not solve the problem. Tell me your version of react, vite? I also tried adding rollup in a new project, which only has ketcher (which I used as an example to reproduce the bug), and it still gives the same error

**Comment by rrodionov91** [comment]
Hi @hellhorse123 
If you can create repository with example we can try to help.

**Comment by hellhorse123** [comment]
> Hi @hellhorse123 If you can create repository with example we can try to help.
This is archive where reproduce the problem:
[ketcher-check.zip](https://github.com/user-attachments/files/17551785/ketcher-check.zip)

However, I don't know how much the above solutions are a crutch, but the following list of steps helped solve this problem:
1) `npm install -D rollup@4.22.0 `(or `yarn add rollup@4.22.0 --dev `)
2) `yarn add vite-plugin-commonjs`
3) in file vite.config:
```
{
plugins: [react(), commonjs()],
define: {
    'process.env': {},
    global: 'global',
  },
  build: {
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },
}
```

4) in file index.html :
```
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + React + TS</title>
         // add script code add to head
    <script>
      var global = window;
    </script>
  </head>
```

5) And in my case, was additional problem with cdn plugin in vite.config. I remove react and react-dom from it and all works now

**Comment by hwknsj** [comment]
I also encountered this same issue, along with some [issues resolving the types for `ketcher-standalone` ](https://github.com/epam/ketcher/issues/5760)that introduced themselves after upgrading to the latest `v2.25.0`.

It would be great to resolve this issue without having to do an elaborate workaround such as the [ones](https://github.com/epam/ketcher/issues/5565#issuecomment-2443223993) @hellhorse123 and @wiscgazf [describe](https://github.com/epam/ketcher/issues/5565#issuecomment-2439192388) (thanks to both of you, by the way!).

@rrodionov91 @samellis – any idea if we can expect a fix for this issue or has it been resolved?

**Comment by LironToledo** [comment]
Im experiencing this exact same issue, and the solution suggested in this thread didn't work for me...

**Comment by chiatzeheng** [comment]
@

> ~Im experiencing this exact same issue, and the solution suggested in this thread didn't work for me...~
> 
> Edit:
> 
> Tried again this morning and adding just this to vite config actually fixed it for me:
> 
> ```
>  build: {
>     commonjsOptions: {
>       transformMixedEsModules: true,
>     },
>   },
> ```

Thank you! This worked for me

**Comment by amcraig** [comment]
For anyone coming after:
This is what worked for me:
```
  build:{
    commonjsOptions: {
      include: [/ketcher/, /node_modules/],
      transformMixedEsModules: true,
    },
    ...
  },
  ...
  optimizeDeps: {
    include: ["ketcher-core", "ketcher-react", "ketcher-standalone"],
  },
```

And then I had to set an Error Boundary around the Editor component to contain other Ketcher issues from bleeding out into the rest of the app.

**Comment by AdolfTorvald** [comment]
> Hi guys! Sorry that I didn't answer before. I'm happy that you found a workaround.
> 
> I will raise this topic on our meetings and propose creating example apps for each modern setup tools and include it into our pipeline to ensure that nothing is broken. Currently we are using only Vite + Webpack example.
> 
> Thank you for feedback!

In the new project, I was able to get everything up and running with these parameters:

1)libs:
```
    "ketcher-react": "^3.7.0",
    "ketcher-standalone": "^3.7.0",
    "react": "^19.1.1",
    "react-dom": "^19.1.1",
    "vite": "^7.1.4",
    "typescript": "^5.9.2",
```

2) vite.config.ts:
```
export default defineConfig({
...
  build: {
    commonjsOptions: {
      include: [/ketcher-standalone/, /ketcher-react/, /raphael/, /node_modules/],
      transformMixedEsModules: true,
    },
  },
...
```

3) new file ketcher.d.ts :
```
/* eslint-disable @typescript-eslint/no-explicit-any */
declare module 'ketcher-standalone' {
  export class StandaloneStructServiceProvider {
    mode: 'standalone';
    createStructService: () => any;
  }
}
interface Window {
  ketcher: any;
}

```

4) component with ketcher:
```
import 'ketcher-react/dist/index.css';

type Ketcher = {
  setMolecule: (data: string) => Promise<void> | void;
  getSmiles: () => Promise<string>;
};

// I tried to use lazy loading for ketcher in modal mode, a small hack. If has different way for @tanstack/react-router, write to me
const Editor = lazy(() =>
  import('ketcher-react').then((module) => ({ default: module['Editor'] })),
);
  useEffect(() => {
    if (isModalOpen && !structServiceProvider) {
      import('ketcher-standalone').then((module) => {
        setStructServiceProvider(new module.StandaloneStructServiceProvider());
      });
    }
  }, [isModalOpen, structServiceProvider]);


...
  const [ketcherInstance, setKetcherInstance] = useState<Ketcher | null>(null);
...
          <Editor
            buttons={{
              sgroup: { hidden: true },
              text: { hidden: true },
            }}
            staticResourcesUrl={process.env.PUBLIC_URL ?? '/'}
            structServiceProvider={structServiceProvider}
            errorHandler={(message: string | Error) => {
              setHasError(true);
              setErrorMessage(typeof message === 'string' ? message : message.toString());
            }}
            onInit={(ketcher: Ketcher) => {
              // сохраняем инстанс один раз
              if (!ketcherInstance) {
                window.ketcher = ketcher;
                setKetcherInstance(ketcher);
                try {
                  ketcher.setMolecule(defaultSmilesValue || '');
                } catch {
                  setHasError(true);
                  setErrorMessage('Ошибка инициализации Ketcher.');
                }
              }
            }}
          />
```

**Comment by aslmbk** [comment]
@hellhorse123 @rrodionov91 is this issue still relevant?

**Comment by yaroslav-b-io** [comment]
@AlexeyGirin  
**SWC panic (width 3 given for non-narrow character)**                                                                     
                                                                                                                                     
  Root cause confirmed: this was a bug in Rollup's internal Rust parser (swc_common crate), not in ketcher's source code. Reproduced 
  the exact repro steps (Vite 5.3.2, @vitejs/plugin-react-swc, ketcher-react@2.24.0) with a current toolchain and the build passes   
  cleanly through the transform phase — no panic. Fixed upstream in rollup@4.22.0 (confirmed by @samellis above). No fix needed on   
  our end; closing this part as resolved-upstream.                                                                                   
                                                                                                                                                                                                                                                                
**Uncaught ReferenceError: require is not defined**                                                                        
                                                                                                                                     
  Root cause: raphael-ext.ts loads Raphael conditionally with typeof window !== 'undefined' ? require('raphael') : undefined — the   
  window check exists to keep this SSR-safe (Raphael touches DOM globals at import time and crashes under Node). But the runtime     
  require() call itself has no browser equivalent, and strict-ESM bundler output (Vite/Rollup) doesn't provide a require global, so  
  it throws and aborts the whole bundle.                                                                                             
                                                                                                                                     
  Fix: added a second guard, typeof require !== 'undefined', so the module degrades to undefined instead of crashing when require    
  isn't available
                                                                                                                                                                                                                                                    
  This stops the hard crash but doesn't by itself make Raphael load under strict-ESM Vite builds — you'll still want the             
  commonjsOptions.transformMixedEsModules + optimizeDeps.include config already shared in this thread (credit @amcraig,              
  @AdolfTorvald) to get a fully working editor. A bundler-agnostic fix (dynamic import()) would require refactoring the render init  
  path to be async — out of scope here, tracked via #7977.
